third_party/nvfuser/csrc/expr_simplifier.cpp

```diff
@@ -133,52 +133,112 @@ std::unique_ptr<debug_print::NoOpLogger> createLogger(Val* value) {
+      std::vector<Bool*> assumptions,
+      bool preserve_error)
       : preserve_error_(preserve_error) {
-    var_info_map_.reserve(variables.size());
     var_order_.reserve(variables.size());
-    set_.reserve(variables.size());
+    var_set_.reserve(variables.size());
+    less_than_.reserve(assumptions.size());
+    less_equal_.reserve(assumptions.size());
     for (const auto& info : variables) {
-      var_order_.emplace_back(info.variable);
-      set_.emplace(info.variable);
-      var_info_map_[info.variable] = info;
+      auto var = info.variable;
+      if (info.is_unrolled_loop_index) {
+        unrolled_loop_index_.insert(var);
+      }
+      var_order_.emplace_back(var);
+      var_set_.emplace(var);
+    }
+    // decompose a && b in assumptions as a and b
+    auto axioms = getAxioms();
+    assumptions.insert(assumptions.end(), axioms.begin(), axioms.end());
+    while (!assumptions.empty()) {
+      auto back = assumptions.back();
+      assumptions.pop_back();
+      auto bop = dynamic_cast<BinaryOp*>(back->definition());
+      if (bop == nullptr || bop->getBinaryOpType() != BinaryOpType::And) {
+        assume(back);
+      } else {
+        assumptions.push_back(bop->lhs()->as<Bool>());
+        assumptions.push_back(bop->rhs()->as<Bool>());
+      }
     }
   }
 
-  // Get the order of variables
-  const std::vector<Val*>& order() const {
+  const std::vector<Val*>& variableOrder() const {
     return var_order_;
   }
 
-  // Get the info of a variable
-  const VarInfo& info(Val* var) const {
-    return var_info_map_.at(var);
+  const std::unordered_set<Val*>& variableSet() const {
+    return var_set_;
+  }
+
+  bool preserveError() const {
+    return preserve_error_;
   }
 
-  // Check if this mapping has information about the given variable
-  bool has(Val* var) const {
-    return set_.count(var);
+  bool isUnrolledLoopIndex(Val* x) const {
+    return unrolled_loop_index_.count(x) > 0;
   }
 
-  // Get the set of variables that has information
-  const std::unordered_set<Val*>& set() const {
-    return set_;
+  const std::vector<std::pair<Val*, Val*>>& getKnownLessThan() const {
+    return less_than_;
   }
 
-  bool preserveError() const {
-    return preserve_error_;
+  const std::vector<std::pair<Val*, Val*>>& getKnownLessEqual() const {
+    return less_equal_;
+  }
+
+ private:
+  void assume(Bool* a) {
+    auto def = a->definition();
+    if (auto bop = dynamic_cast<BinaryOp*>(def)) {
+      switch (bop->getBinaryOpType()) {
+        case BinaryOpType::LT:
+          less_than_.emplace_back(bop->lhs(), bop->rhs());
+          break;
+        case BinaryOpType::LE:
+          less_equal_.emplace_back(bop->lhs(), bop->rhs());
+          break;
+        case BinaryOpType::GT:
+          less_than_.emplace_back(bop->rhs(), bop->lhs());
+          break;
+        case BinaryOpType::GE:
+          less_equal_.emplace_back(bop->rhs(), bop->lhs());
+          break;
+        default:
+          break;
```

Comment:
Should we assert here? Looks like, if no other op shows up here, we are just ignoring it, which is fine in terms of correctness, but having unsupported ops seems like a red flag to me.

Reply:
resolved in https://github.com/NVIDIA/Fuser/pull/33